src/test/java/com/pusher/rest/crypto/CryptoUtilTest.java

```diff
@@ -0,0 +1,78 @@
+package com.pusher.rest.crypto;
+
+import com.pusher.rest.data.EncryptedMessage;
+import org.junit.jupiter.api.Test;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class CryptoUtilTest {
+
+    final String encryptedChannel = "private-encrypted-test";
+    final String testMessage = "{\"message\": \"Hello, world!\"}";
+    final byte[] testData = testMessage.getBytes(StandardCharsets.UTF_8);
+    final String base64EncodedEncryptionMasterKey = "VGhlIDMyIGNoYXJzIGxvbmcgZW5jcnlwdGlvbiBrZXk=";
+    final CryptoUtil crypto = new CryptoUtil(base64EncodedEncryptionMasterKey);
+
+    @Test
+    void parseEncryptionMasterKeyWithInvalidKey() {
+        assertThrows(IllegalArgumentException.class, () -> new CryptoUtil(null));
+
+        assertThrows(IllegalArgumentException.class, () -> new CryptoUtil(""));
+
+        assertThrows(IllegalArgumentException.class, () -> new CryptoUtil("TG9yZW0gSXBzdW0ga"));
+    }
+
+    @Test
+    void encrypt() {
+        final EncryptedMessage encryptedMessage = crypto.encrypt(encryptedChannel, testData);
+
+        assertNotNull(encryptedMessage.getNonce());
+        assertNotNull(encryptedMessage.getCiphertext());
+
+        assertNotEquals(testMessage, encryptedMessage.getCiphertext());
+        assertNotEquals(testData, Base64.getDecoder().decode(encryptedMessage.getCiphertext()));
+    }
+
+    @Test
+    void encryptionRandomness() {
+        final EncryptedMessage m1 = crypto.encrypt(encryptedChannel, testData);
+        final EncryptedMessage m2 = crypto.encrypt(encryptedChannel, testData);
+
+        assertNotEquals(m1.getNonce(), m2.getNonce());
+        assertNotEquals(m1.getCiphertext(), m2.getCiphertext());
+    }
+
+    @Test
+    void encryptionMoreRandomnessCheck() {
+        final int expectedSize = 10000;
+
+        Set<String> set = new LinkedHashSet<>();
+        for (int i = 1; i <= expectedSize; i++) {
+            final EncryptedMessage m = crypto.encrypt(encryptedChannel, testData);
+
+            set.add(m.getNonce().concat(m.getCiphertext()));
+        }
+
+        assertEquals(expectedSize, set.size());
+    }
```

Comment:
The idea is to encrypt same payload `expectedSize` times and add them to a Set and then count if we received `expectedSize` items in the Set or not. Since the HashSet drops any duplicate value, we can check if all generated values were unique. 

Now that I read it for second time, I agree it could be more intuitive. I could rewrite it like below and maybe also add a message to the assertion.

```java
final int expectedDistinctItems = 10000;
List<String> items = new ArrayList<>();

for (int i = 1; i <= expectedDistinctItems; i++) {
    final EncryptedMessage m = crypto.encrypt(encryptedChannel, testData);

    items.add(m.getNonce().concat(m.getCiphertext()));
}

assertEquals(expectedDistinctItems, items.stream().distinct().count());
```

Reply:
Ah, that makes sense. It might be worth a comment in the code to explain that duplicate values are dropped, hence the check on length is effectively a check for no duplicates. No need to change the code itself (although you can if you want!)